debug/testlib.py

```diff
@@ -158,6 +157,31 @@ def __init__(self, cmd=None, config=None, debug=False):
                 messaged = True
                 print "Waiting for OpenOCD to examine RISCV core..."
 
+        self.port = self._get_gdb_server_port()
+
+    def _get_gdb_server_port(self):
+        """Get port that OpenOCD's gdb server is listening on."""
+        MAX_ATTEMPTS = 5
+        PORT_REGEX = re.compile(r'(?P<port>\d+) \(LISTEN\)')
+        for _ in range(MAX_ATTEMPTS):
+            with open(os.devnull, 'w') as devnull:
+                output = subprocess.check_output([
+                    'lsof',
+                    '-a',  # Take the AND of the following selectors
+                    '-p{}'.format(self.process.pid),  # Filter on PID
+                    '-iTCP',  # Filter only TCP sockets
+                ], stderr=devnull)
+            matches = list(PORT_REGEX.finditer(output))
+            if len(matches) > 1:
+                raise Exception(
+                    "OpenOCD listening on multiple ports. Cannot uniquely "
+                    "identify gdb server port.")
+            elif matches:
+                [match] = matches
+                return int(match.group('port'))
+            time.sleep(1)
```

Comment:
Sounds good, reduced the sleep time to 0.1s, and increased the max attempts to 50. Empirically, it looks like it took 6 sleeps (~600ms) for me for it to actually pick up the port, so it looks like it actually is a bit slow to start up the server, since I think it actually requires waiting long enough for the RTL simulator to connect to the JTAG interface.